File: packages/android/core-domain/src/test/kotlin/app/notes/domain/UserTest.kt
package app.notes.domain

import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertThrows
import org.junit.jupiter.api.Test

class UserTest {
    private val id = UserId("550e8400-e29b-41d4-a716-446655440000")

    private val defaultPrefs = AIPreferences(
        chatProvider = AIProviderId.Ollama,
        chatModel = "qwen2.5:7b",
        tagProvider = AIProviderId.Ollama,
        tagModel = "llama3.2:3b",
        embeddingProvider = AIProviderId.Ollama,
        embeddingModel = "nomic-embed-text",
    )

    private fun base(
        displayName: String = "Vgaka",
        storage: StorageProviderId? = StorageProviderId.GoogleDrive,
        calendar: CalendarProviderId? = CalendarProviderId.GoogleCalendar,
        prefs: AIPreferences = defaultPrefs,
    ) = User(id, displayName, storage, calendar, prefs, analyticsOptOut = false)

    @Test
    fun `accepts a valid user`() {
        assertEquals("Vgaka", base().displayName)
    }

    @Test
    fun `rejects a blank display name`() {
        assertThrows(IllegalArgumentException::class.java) { base(displayName = "") }
    }

    @Test
    fun `rejects empty model names in AIPreferences`() {
        assertThrows(IllegalArgumentException::class.java) {
            AIPreferences(
                chatProvider = AIProviderId.Ollama,
                chatModel = "",
                tagProvider = AIProviderId.Ollama,
                tagModel = "llama3.2:3b",
                embeddingProvider = AIProviderId.Ollama,
                embeddingModel = "nomic-embed-text",
            )
        }
    }
}

File: packages/domain/src/__tests__/user.test.ts
import { describe, expect, it } from "vitest";
import { makeUser, type AIPreferences, type User } from "../user.js";
import { UserId } from "../ids.js";

const ID = UserId("550e8400-e29b-41d4-a716-446655440000");

const DEFAULT_AI: AIPreferences = {
  chatProvider: "OLLAMA",
  chatModel: "qwen2.5:7b",
  tagProvider: "OLLAMA",
  tagModel: "llama3.2:3b",
  embeddingProvider: "OLLAMA",
  embeddingModel: "nomic-embed-text",
};

function base(overrides: Partial<User> = {}): User {
  return {
    userId: ID,
    displayName: "Vgaka",
    storageProvider: "GOOGLE_DRIVE",
    calendarProvider: "GOOGLE_CALENDAR",
    aiPrefs: DEFAULT_AI,
    analyticsOptOut: false,
    ...overrides,
  };
}

describe("makeUser", () => {
  it("accepts a valid user", () => {
    expect(makeUser(base()).displayName).toBe("Vgaka");
  });

  it("rejects a blank display name", () => {
    expect(() => makeUser(base({ displayName: "" }))).toThrow(/displayName/);
  });

  it("rejects an unknown storage provider", () => {
    expect(() => makeUser(base({ storageProvider: "DROPBOX" as User["storageProvider"] }))).toThrow(
      /storageProvider/,
    );
  });

  it("rejects an unknown calendar provider", () => {
    expect(() =>
      makeUser(base({ calendarProvider: "FANTASTICAL" as User["calendarProvider"] })),
    ).toThrow(/calendarProvider/);
  });

  it("rejects empty model names", () => {
    expect(() => makeUser(base({ aiPrefs: { ...DEFAULT_AI, chatModel: "" } }))).toThrow(
      /chatModel/,
    );
  });
});
